common/transport/amqp/src/amqp.ts

```diff
@@ -145,44 +235,190 @@ export class Amqp {
+            // a good session set up.  This way we won't wait around for a session end that probably won't come.
+            //
+            this._sessionCloseOccurred = true;
+            this._sessionCallback = undefined;
+            this._fsm.transition('disconnecting', callback, context.connection.error);
+          },
+          'disconnected': (context: EventContext) => {
+            debug('In the disconnect handler for the connecting session state with error of: ' + this._getErrorName(context.connection.error));
+            let callback = this._sessionCallback;
+            this._sessionCallback = undefined;
+            manageConnectionHandlers('removeListener');
+            this._fsm.transition('disconnected', callback, new errors.NotConnectedError('rhea: connection disconnected'));
+          },
+          amqpError: (err) => {
+            this._fsm.transition('disconnecting', null, err);
           },
-          amqpError: (err) => this._fsm.transition('disconnecting', null, err),
-          amqpDisconnected: () => debug('ignoring disconnected event while connecting'),
           '*': () => this._fsm.deferUntilTransition()
         },
         connected: {
           _onEnter: (connectCallback, result) => {
             /*Codes_SRS_NODE_COMMON_AMQP_16_002: [The `connect` method shall establish a connection with the IoT hub instance and if given as argument call the `done` callback with a null error object in the case of success and a `results.Connected` object.]*/
             this._safeCallback(connectCallback, null, new results.Connected(result));
           },
-          amqpError: (err) => this._fsm.transition('disconnecting', null, err),
-          amqpDisconnected: () => this._fsm.transition('disconnected', undefined, new errors.NotConnectedError('amqp10: connection closed')),
+          'session_error': (context: EventContext) => {
+            debug('In the session_error handler for the connected state with error of: ' + this._getErrorName(context.session.error));
+            debug('Deferring transition until the session_close event occurs.');
+            this._indicatedSessionError = context.session.error;
+          },
+          'session_close': (context: EventContext) => {
+            debug('In the session_close handler for the connected state with indicated error of: ' + this._getErrorName(this._indicatedSessionError));
+            debug('Transition to disconnecting');
+            let err = this._indicatedSessionError;
+            this._indicatedSessionError = undefined;
+            this._sessionCloseOccurred = true;
+            this._fsm.transition('disconnecting', null, err);
+          },
+          'connection_error': (context: EventContext) => {
+            debug('In the connection_error handler for the connected state with error of: ' + this._getErrorName(context.connection.error));
+            debug('Deferring transition until the connection_close occurs.');
```

Comment:
> debug('Deferring transition until the connection_close occurs.'); [](start = 11, length = 66)

vague uncomfortable feeling.  It looks like you're assuming that connection_close will maybe always happen after connection_error.  Is this a safe assumption?  Is the connection in some sort of "broken and waiting to close" state here?  Or is the state more vague than that?

Reply:
Connection close ALWAYS will be called after the connection error.  Note though that connection_close is NOT always preceded by connection_error.  This statement is also true for all *_error and *_close events used throughout rhea.  Other than the general pain of the event handling, I don't consider this a deficiency.

This actually makes fairly reasonable sense if you examine how errors are presented as specified by the AMQP performative specification.  The way that errors show up is VIA a performative that causes the *_close events. 

---
In reply to: [209388831](https://github.com/Azure/azure-iot-sdk-node/pull/344#discussion_r209388831) [](ancestors = 209388831)